### PR TITLE
Add `ExpiryCaveat`

### LINKED ISSUE #6255: Add caveats to macaroons for expiration (time) and version
**What's the problem this feature will solve?**

This will allow further attenuating the permissions granted by an API key

**Describe the solution you'd like**

Addition of two addition types of caveat, project version (for uploads) and time (expiry).

### LINKED ISSUE #10792: Update Macaroon claims/constraints for OIDC token minting
We're going to want to mint short-lived Warehouse API tokens (which are Macaroons) based on incoming OIDC JWTs.

As part of the "short-lived" part, we'll need need to make sure the tokens support some kind of expiration claim.

### PR DESCRIPTION
**Pull request by woodruffw**
Closes #10792.

Closes #6255.

### CODE REVIEW

**Comment by woodruffw** [comment]
This needs more testing, but the basic idea:

* All current macaroons continue to function as expected, and nothing about the construction of normal macaroons (i.e., ones minted by users manually through the PyPI website) is changed.

* In the future, macaroons minted via the OIDC JWT consumption route will include an additional expiry caveat, with two fields: `nbf` to indicate that the macaroon must not be used before the given timestamp, and `exp` to indicate when the macaroon ceases to be valid.

File: warehouse/macaroons/services.py
Comment (woodruffw) [inline warehouse/macaroons/services.py:142]: Highlighting this as a bit of a hack: I'm not really happy with this trick, but the alternative would be a `create_macaroon` signature like this:

```python
def create_macaroon(self, location, user_id, description, caveats, permissions):
  ...
```

...where `caveats[N] == permissions`. That feels roughly as bad, so I figured reducing the parameter deduplication was a reasonable tradeoff. But I'm very open to better ideas.

File: warehouse/manage/views.py
Comment (woodruffw) [inline warehouse/manage/views.py:885]: Highlighting this as another part I'm not satisfied with: we subsequently assume that `macaroon_caveats[0]` is the permissions caveat, e.g. when pretty-printing events. That's unlikely to change, but it's a piece of nasty implicit state here.

Comment (woodruffw) [reply warehouse/manage/views.py:885]: (The other option is to embed `form.validated_scope` into the event instead of the entire `macaroon_caveats` list. That would avoid the implicit list position.)

File: warehouse/macaroons/services.py
Comment (woodruffw) [inline warehouse/macaroons/services.py:143]: N.B.: I don't think there's any actual coverage missing here; I think `coverage` is getting tripped up by the `StopIteration`.

**Comment by woodruffw** [comment]
Marking as ready for review since this is in a stable state, but see comments above about the current approach.

File: warehouse/manage/views.py
Comment (di) [reply warehouse/manage/views.py:885]: Or maybe we should update the UI to include all caveats?

**Review by di — COMMENTED** [review]
This looks good, but I didn't give it a super deep review. I had one high level question: how will this be backwards-compatible for macaroons created before this change? The migration is just dropping one column and adding a new one without doing a data migration.

**Comment by woodruffw** [comment]
> I had one high level question: how will this be backwards-compatible for macaroons created before this change?

Yeah, this was an oversight on my part. This will break the UI for pre-existing macaroons, at least in the form they're recorded in the DB.

The migration should probably just rename the column rather than dropping it and inserting a new one, since interior data is the same. I'll fix that now.

File: warehouse/manage/views.py
Comment (woodruffw) [reply warehouse/manage/views.py:885]: Alright, I've added a `caveat_detail` macro that handles both caveat types, removing the need for the implicit array positioning here.

The changes introduced with that macro don't currently manifest in the UI since we don't currently mint any API tokens with the new expiry caveat, but they'll show up once I begin the route work.

**Comment by woodruffw** [comment]
Blocked on: #11157.

**Comment by di** [comment]
#11157 has been merged!

**Comment by woodruffw** [comment]
Excellent! I'll de-conflict here.

**Comment by woodruffw** [comment]
N.B.: [1f65cff](https://github.com/pypa/warehouse/pull/11122/commits/1f65cff664bef00de542e7a1ea4a3eba2172ad52) is a naive pick of the original changeset, so it isn't ready for review yet. I'll do another pass over it later today.

**Comment by woodruffw** [comment]
I also confirmed locally that this does not break verification of current tokens (i.e., ones missing an expiry caveat).

File: warehouse/macaroons/caveats.py
Comment (di) [inline warehouse/macaroons/caveats.py:106]: I was expecting to see a test change as a result of this, but it doesn't look like we're testing the `Verifier` class that well. Could you add a test that captures the change in behavior that this line introduces?

Comment (woodruffw) [reply warehouse/macaroons/caveats.py:111]: This was a great call -- adding these tests ([3e60c4b](https://github.com/pypa/warehouse/pull/11122/commits/3e60c4b51e856a966a4a79259d57cf4ad27eed53)) caught a bug that would have caused us to unconditionally fail all expiry caveat verifications. I've refactored the code to avoid exception-driven failures in caveats going forwards, since that's the root of the problem.

File: warehouse/macaroons/caveats.py
Comment (woodruffw) [inline warehouse/macaroons/caveats.py:32]: This change will make #9018 much easier to implement -- we just need to plumb the `failure_reason` through the verifier and service layers.

**Review by ewdurbin — APPROVED** [review]

File: warehouse/macaroons/caveats.py
Comment (di) [inline warehouse/macaroons/caveats.py:37]: I'm not sure I love this pattern, for two reasons:

- It's not hard to imagine someone calling `self._fail(...)` (and expecting that call to change the flow) instead of `return self._fail(...)`, accidentally returning the implicit `None` instead. 
- The fact that `self._fail` essentially wraps setting state and just always returns `False` seems unnecessary.

I think I'd prefer sticking with an exception-driven control flow but I'm not sure I understand why you opted to move aware from it here.

Comment (woodruffw) [reply warehouse/macaroons/caveats.py:37]: I'm also not in love with this pattern, but the original (exception-driven) control flow was actually masking a bug in caveat verification (one that we never experienced, since we've only ever had one caveat per macaroon before).

Using exceptions here was actually causing us to prematurely fail our caveat verification steps, since `pymacaroons` runs each verifier once per caveat body. This meant that the `V1Caveat` would raise as soon as it was passed an `ExpiryCaveat` body, failing the entire verification phase rather than returning control to the next verifier.

IMO we need _some_ kind of state handler to make the error messages returned for macaroons more useful (per #9018), although it definitely doesn't need to be this particular `_fail` thing I just made up :slightly_smiling_face:

Comment (di) [reply warehouse/macaroons/caveats.py:37]: Got it, that makes sense.

As a compromise, can we just do something like this instead::

```suggestion
        self.failure_reason = "programming error"
        return False
```

and add some static typing to ensure that `verify` always returns a boolean?

Comment (woodruffw) [reply warehouse/macaroons/caveats.py:37]: Yep, SGTM. I'll make these changes now.

Comment (woodruffw) [reply warehouse/macaroons/caveats.py]: [e0d1709](https://github.com/pypa/warehouse/pull/11122/commits/e0d1709a1898b9c1934f7c9860d6ed2c5daa468b)

**Review by di — APPROVED** [review]

**Comment by fschulze** [comment]
What is the deployment policy of pypi.org? Is everything on main automatically deployed? If not, where can I see what version is deployed? In other words, how can I see that changes like this are available on pypi.org?

**Comment by di** [comment]
@fschulze Everything is automatically deployed within 5-10 minutes after merge.

**Comment by woodruffw** [comment]
But also, note that these changes aren't user-facing yet! We don't currently allow users to create API tokens with expiry restrictions.

**Pull request merged**
